File: app/src/main/java/com/example/pam_project/features/lists/list/ListAdapter.java
package com.example.pam_project.features.lists.list;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.pam_project.R;
import com.example.pam_project.features.categories.list.CategoryInformation;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class ListAdapter extends RecyclerView.Adapter<ListViewHolder> implements Filterable {
    private static List<Integer> filterSelections;
    private static int sortIndex = 0;

    private final List<ListInformation> dataSet;
    private final List<ListInformation> hiddenItems;
    private final List<ListInformation> completeDataset;
    private final List<ListInformation> previousSearchDataset;
    private List<ListInformation> searchItems;
    private final Map<Long, CategoryInformation> categoriesWithIds;
    private List<CategoryInformation> categories;
    private OnListClickedListener listener;

    public ListAdapter() {
        this.dataSet = new ArrayList<>();
        this.categoriesWithIds = new HashMap<>();
        this.categories = new ArrayList<>(categoriesWithIds.values());
        this.hiddenItems = new ArrayList<>();
        this.completeDataset = new ArrayList<>();
        this.previousSearchDataset = new ArrayList<>();
        this.searchItems = new ArrayList<>();
    }

    public void update(final List<ListInformation> newDataSet) {
        this.dataSet.clear();

        if (newDataSet != null)
            dataSet.addAll(newDataSet);

        setFilterSelections(filterSelections); // sort and filter accordingly
        notifyDataSetChanged();
    }

    public void updateCategories(final List<CategoryInformation> newCategories) {
        categoriesWithIds.clear();
        if (newCategories != null) {
            for (CategoryInformation category : newCategories) {
                categoriesWithIds.put(category.getId(), category);
            }
        }
        categories = new ArrayList<>(categoriesWithIds.values());

        for (ListInformation list : dataSet) {
            list.setCategory(categoriesWithIds.get(list.getCategoryId()));
        }

        if (filterSelections == null)
            return;

        // reset category selections if one category has been deleted and was selected
        for (int i = 0; i < filterSelections.size(); i++) {
            if (!categoriesWithIds.containsKey(Long.valueOf(filterSelections.get(i)))) {
                filterSelections = null;
                return;
            }
        }
    }

    public void setOnClickedListener(final OnListClickedListener listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public ListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.list_view_holder, parent, false);
        return new ListViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ListViewHolder holder, int position) {
        holder.bind(dataSet.get(position));
        holder.setOnClickListener(listener);
    }

    @Override
    public int getItemCount() {
        return dataSet.size();
    }

    public int getSortIndex() {
        return sortIndex;
    }

    public void setSortIndex(int sortIndex) {
        ListAdapter.sortIndex = sortIndex;
        sort();

        notifyDataSetChanged();
    }

    public List<Integer> getFilterSelections() {
        return filterSelections;
    }

    public List<CategoryInformation> getCategories() {
        return Collections.unmodifiableList(categories);
    }

    public void setCompleteDataset(final List<ListInformation> dataset){
        completeDataset.clear();
        completeDataset.addAll(dataset);
    }

    public void setPreviousSearchDataset(){
        previousSearchDataset.clear();
        previousSearchDataset.addAll(dataSet);
    }

    public List<ListInformation> getPreviousSearchDataset(){
        return new ArrayList<>(previousSearchDataset);
    }

    public void setFilterSelections(final List<Integer> newFilterSelections) {
        if (newFilterSelections == null) {
            sort();
            return;
        }

        final List<Long> selectedCategoriesIds = new ArrayList<>(
                newFilterSelections.size());
        // get ids from selected categories
        for (Integer index : newFilterSelections)
            selectedCategoriesIds.add(categories.get(index).getId());

        // hide items not in selected categories and remove them from dataSet
        moveToList(dataSet, hiddenItems, (id) -> !selectedCategoriesIds.contains(id));

        // show items in selected categories and remove them from hidden items
        moveToList(hiddenItems, dataSet, selectedCategoriesIds::contains);


        removeDuplicates(dataSet);
        removeDuplicates(hiddenItems);
        sort();

        notifyDataSetChanged();
        setPreviousSearchDataset();
        ListAdapter.filterSelections = newFilterSelections;
    }

    private void sort() {
        Collections.sort(dataSet, ListInformation.getComparator(sortIndex));
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                String charString = charSequence.toString();
                if (charString.isEmpty()) {
                    searchItems = completeDataset;
                } else {
                    List<ListInformation> filteredList = new ArrayList<>();
                    for (ListInformation listInformation : completeDataset) {
                        if (listInformation.getTitle().toLowerCase().contains(charString.toLowerCase())) {
                            filteredList.add(listInformation);
                        }
                    }

                    searchItems = filteredList;
                }

                FilterResults filterResults = new FilterResults();
                filterResults.values = searchItems;
                return filterResults;
            }

            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                List<?> result = (List<?>) filterResults.values;
                ArrayList<ListInformation> tempList = new ArrayList<>();
                for (Object object : result) {
                    if (object instanceof ListInformation) {
                        tempList.add((ListInformation) object); // <-- add to temp
                    }
                }

                dataSet.clear();
                dataSet.addAll(tempList);
                notifyDataSetChanged();
            }
        };
    }

    @FunctionalInterface
    interface ListCheck {
        boolean condition(long id);
    }

    private void moveToList(List<ListInformation> from, List<ListInformation> to,
                           ListCheck function) {
        // show items in dataSet or hidden categories
        for (ListInformation item : from) {
            if (function.condition(item.getCategoryId())) {
                to.add(item);
            }
        }
        // remove shown items from hidden elements or dataSet
        for (ListInformation item : to)
            from.remove(item);
    }

    private void removeDuplicates(List<ListInformation> list){
        List<ListInformation> listNoDuplicates = new ArrayList<>();
        Set<Long> idSet = new HashSet<>();

        for(ListInformation listInformation: list){
            if(! idSet.contains(listInformation.getId())){
                idSet.add(listInformation.getId());
                listNoDuplicates.add(listInformation);
            }
        }

        list.clear();
        list.addAll(listNoDuplicates);
    }
}

File: app/src/main/java/com/example/pam_project/features/lists/list/ListAdapter.kt
package com.example.pam_project.features.lists.list

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.*
import androidx.recyclerview.widget.RecyclerView
import com.example.pam_project.R
import com.example.pam_project.features.categories.list.CategoryInformation
import java.util.*

class ListAdapter : RecyclerView.Adapter<ListViewHolder>(), Filterable {
    private val dataSet: MutableList<ListInformation?>
    private val hiddenItems: MutableList<ListInformation?>
    private val completeDataset: MutableList<ListInformation?>
    val previousSearchDataset: MutableList<ListInformation?>
    private var searchItems: List<ListInformation?>
    private val categoriesWithIds: MutableMap<Long, CategoryInformation?>
    private var listener: OnListClickedListener? = null
    var categories: List<CategoryInformation?>

    var sortIndex: Int
        get() = Companion.sortIndex
        set(sortIndex) {
            Companion.sortIndex = sortIndex
            sort()
            notifyDataSetChanged()
        }

    fun update(newDataSet: List<ListInformation?>?) {
        dataSet.clear()
        if (newDataSet != null) dataSet.addAll(newDataSet)
        filterSelections = Companion.filterSelections // sort and filter accordingly
        notifyDataSetChanged()
    }

    fun updateCategories(newCategories: List<CategoryInformation?>?) {
        categoriesWithIds.clear()
        if (newCategories != null) {
            for (category in newCategories) {
                categoriesWithIds[category!!.id] = category
            }
        }
        categories = ArrayList(categoriesWithIds.values)
        for (list in dataSet) {
            list?.category = categoriesWithIds[list?.categoryId]
        }
        if (Companion.filterSelections == null) return

        // reset category selections if one category has been deleted and was selected
        for (i in Companion.filterSelections!!.indices) {
            if (!categoriesWithIds.containsKey(filterSelections!![i].toLong())) {
                Companion.filterSelections = null
                return
            }
        }
    }

    fun setOnClickedListener(listener: OnListClickedListener?) {
        this.listener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.list_view_holder, parent, false)
        return ListViewHolder(view)
    }

    override fun onBindViewHolder(holder: ListViewHolder, position: Int) {
        holder.bind(dataSet[position])
        holder.setOnClickListener(listener)
    }

    override fun getItemCount(): Int {
        return dataSet.size
    }

    // get ids from selected categories

    // hide items not in selected categories and remove them from dataSet

    // show items in selected categories and remove them from hidden items
    var filterSelections: MutableList<Int>?
        get() = Companion.filterSelections
        set(newFilterSelections) {
            if (newFilterSelections == null) {
                sort()
                return
            }
            val selectedCategoriesIds: MutableList<Long> = ArrayList(
                    newFilterSelections.size)
            // get ids from selected categories
            for (index in newFilterSelections) selectedCategoriesIds.add(categories[index]?.id!!)

            // hide items not in selected categories and remove them from dataSet
            moveToList(dataSet, hiddenItems) { id: Long -> !selectedCategoriesIds.contains(id) }

            // show items in selected categories and remove them from hidden items
            moveToList(hiddenItems, dataSet) { o: Long -> selectedCategoriesIds.contains(o) }
            removeDuplicates(dataSet)
            removeDuplicates(hiddenItems)
            sort()
            notifyDataSetChanged()
            setPreviousSearchDataset()
            Companion.filterSelections = newFilterSelections
        }

    fun setCompleteDataset(dataset: List<ListInformation?>?) {
        completeDataset.clear()
        completeDataset.addAll(dataset!!)
    }

    fun setPreviousSearchDataset() {
        previousSearchDataset.clear()
        previousSearchDataset.addAll(dataSet)
    }

    private fun sort() {
        Collections.sort(dataSet, ListInformation.getComparator(Companion.sortIndex))
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence): FilterResults {
                val charString = charSequence.toString()
                searchItems = if (charString.isEmpty()) {
                    completeDataset
                } else {
                    val filteredList: MutableList<ListInformation?> = ArrayList()
                    for (listInformation in completeDataset) {
                        if (listInformation?.title?.lowercase()?.contains(charString.lowercase())!!) {
                            filteredList.add(listInformation)
                        }
                    }
                    filteredList
                }
                val filterResults = FilterResults()
                filterResults.values = searchItems
                return filterResults
            }

            override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
                val result = filterResults.values as List<*>
                val tempList = ArrayList<ListInformation?>()
                for (`object` in result) {
                    if (`object` is ListInformation) {
                        tempList.add(`object` as ListInformation?) // <-- add to temp
                    }
                }
                dataSet.clear()
                dataSet.addAll(tempList)
                notifyDataSetChanged()
            }
        }
    }

    internal fun interface ListCheck {
        fun condition(id: Long): Boolean
    }

    private fun moveToList(from: MutableList<ListInformation?>, to: MutableList<ListInformation?>,
                           function: ListCheck) {
        // show items in dataSet or hidden categories
        for (item in from) {
            if (function.condition(item?.categoryId!!)) {
                to.add(item)
            }
        }
        // remove shown items from hidden elements or dataSet
        for (item in to) from.remove(item)
    }

    private fun removeDuplicates(list: MutableList<ListInformation?>) {
        val listNoDuplicates: MutableList<ListInformation?> = ArrayList()
        val idSet: MutableSet<Long> = HashSet()
        for (listInformation in list) {
            if (!idSet.contains(listInformation?.id)) {
                idSet.add(listInformation?.id!!)
                listNoDuplicates.add(listInformation)
            }
        }
        list.clear()
        list.addAll(listNoDuplicates)
    }

    companion object {
        private var filterSelections: MutableList<Int>? = null
        private var sortIndex = 0
    }

    init {
        dataSet = ArrayList()
        categoriesWithIds = HashMap()
        categories = ArrayList(categoriesWithIds.values)
        hiddenItems = ArrayList()
        completeDataset = ArrayList()
        previousSearchDataset = ArrayList()
        searchItems = ArrayList()
    }
}